### DISCUSSION TITLE
Exporting from justwatch

### DISCUSSION THREAD

**Discussion by bagajohny** [DiegoFleitas/letterboxd-movie-justwatch#30] [Q&A] [answered]
Hi @DiegoFleitas 

As per your suggestion I have tried github codespaces and I am facing some issue. Please check the screenshot below,

![image](https://github.com/DiegoFleitas/letterboxd-movie-justwatch/assets/101859073/cc634bde-679b-429d-94d9-0c7664c0e45c)

What could be the reason? I believe the format of the urls is correct. I will attach the txt file for your reference in this post. Please help.
[urls.txt](https://github.com/DiegoFleitas/letterboxd-movie-justwatch/files/13758590/urls.txt)

**Comment by DiegoFleitas** [comment, marked as answer]
It worked for me so its likely to be something wrong with your urls.txt file, if I had to guess its the line endings, I see your file uses CRLF but Unix expects LF (like mine) and sometimes that can break stuff. You can ask ChatGPT about "CRLF vs LF line endings and curl (3): URL using bad/illegal format or missing URL" its I'm sure It can give you a better explanation than I ever could

![imagen](https://github.com/DiegoFleitas/letterboxd-movie-justwatch/assets/15369935/db62c85e-646f-499f-b0bb-a7f924c4e869)

**Reply by bagajohny** [reply]
You are correct. I changed the line endings to LF and now the code is working. Thanks to you i was able to download all the html pages!

![image](https://github.com/DiegoFleitas/letterboxd-movie-justwatch/assets/101859073/e352fb55-b216-4622-890d-9c8358d968e8)

I have downloaded all the pages in the URLs folder. from here how do I run the python script to extract the data to a json file?

**Reply by DiegoFleitas** [reply]
Keep using the codespace, this is by far the best way to continue.
This is where the guide has some issues on my opinion

1. the script that downloads the html doesn't do it inside a folder by default
2. the downloaded files don't have the correct extension either
3. it assumes the required python dependencies have been installed already (not our case)

You can run this command on the codespace to move all the html files that have no extension inside a media folder
`mkdir -p media && find . -maxdepth 1 -type f ! -name "*.*" -exec mv {} media/ \;` (or if you have all the files inside URLs already just rename that folder to media)
then this other command to begin with the python dependency installation
`echo -e "lxml==4.6.3\nrequests==2.25.1\nBeautifulSoup4==4.9.3" > requirements.txt`
install the dependencies
`pip install -r requirements.txt`
last, run the script (& pray it works 😉 )
`python import.py media`

It most likely won't. I don't think it is actively being mantained either but understanding what it's trying to do is crucial to know what needs fixing and continue going forward

**Reply by bagajohny** [reply]
> the downloaded files don't have the correct extension either

Yes. I thought it was some issue from my side. So I later changed the extensions of all files to html.

I tried as you said. I am getting this,

![image](https://github.com/DiegoFleitas/letterboxd-movie-justwatch/assets/101859073/e280a588-117c-4687-a77a-da4323fa0278)

**Reply by DiegoFleitas** [reply]
Yup, that's the script failing as predicted. Options are either debug & fix that script yourself or create a different one, both require to know how to code & I won't be able to give you a step by step process for that. 
Since I'm invested on your issue I might code a solution myself with a bookmarklet (JS that runs in the browser & runs on demand, like clicking a bookmark) like I originally proposed when you explained what you were doing but that might take me a while

**Reply by bagajohny** [reply]
I see. I have a few friends who can code so I will show it to them to see if they can fix it. Chances are low but I will at least try. If I am able to get it working I will definitely ping you.
If you have a different way then please work on it. I am not at all pressed for time so I can wait. Simkl has many different ways to import data. So you can try whichever is easiest for you from below.

![image](https://github.com/DiegoFleitas/letterboxd-movie-justwatch/assets/101859073/aad9f272-005a-47c3-93c2-9de83ca8dcf7)

**Reply by DiegoFleitas** [reply]
Simkl seems to want to charge for interfacing with advanced importing methods so AFAIK we are stuck with using .csv

Here's my crack at it https://diegofleitas.github.io/bookmarklet-collection/ I called it "justwatch watchlist exporter" & there is some documentation here https://github.com/DiegoFleitas/bookmarklet-collection/tree/main/justwatch-watchlist-exporter
I warn you it will take time to run if you have a lenghty list (10s or so per movie), I also didn't test all cases so it's bound to have some bugs too

**Reply by bagajohny** [reply]
I want to ask something. Why cant we use the downloaded html pages with your method? Is it really required that we use the website in browser? I mean will not be easier and quicker to use the downloaded html pages since they have all the same data as website page?

And about the scrpit, do we need to keep the browser tab in focus all the while the script is running or can we minimise the browser? Because after opening the 2nd or 3rd popup nothing happens and I think the script gets stuck. Can we not scroll the opened popup page at all while the script is running?

**Reply by DiegoFleitas** [reply]
> I want to ask something. Why cant we use the downloaded html pages with your method? Is it really required that we use the website in browser? I mean will not be easier and quicker to use the downloaded html pages since they have all the same data as website page?

- I would rather write code on javascript than python, the language the original script was written into
- You have all the data in the browser already, the script is deliberately slow for the sake of simplicity (see https://github.com/DiegoFleitas/bookmarklet-collection/blob/32dde624eda0be10f83816f6c3f89f8083f6bafe/justwatch-watchlist-exporter/index.js#L6) to not have to deal with race conditions, parallel execution or parsing html (usually not a good idea)

> And about the scrpit, do we need to keep the browser tab in focus all the while the script is running or can we minimise the browser? Because after opening the 2nd or 3rd popup nothing happens and I think the script gets stuck. Can we not scroll the opened popup page at all while the script is running?

- You don't need to interact with the page but I belive your browser window will regain focus every time a new tab is opened so the smart thing to do is wait for it to finish running. If successful it should close all the tabs it opened and download a CSV
- About getting stuck you are likely running into issues trying to export TV Shows, which neither the original script nor my script have implemented code for. Both can only handle movies at the moment (this is likely due to movies and tv shows being their own different thing when it comes to tracking media)

**Reply by DiegoFleitas** [reply]
Btw I updated the script, it won't open new tabs now. But you'll need to "reinstall" meaning delete the bookmark from your browser and dragging it back to the bookmark bar from my web, so you get the latest code https://diegofleitas.github.io/bookmarklet-collection/

**Reply by bagajohny** [reply]
I see. I didn't know that it will not work with tv shows. I will use the updated script but just to be sure, the updated script also works for only movies right?

**Reply by bagajohny** [reply]
And also about the localstorage save, where is the csv file saved actually? Will it be inside my downloads folder of firefox or somewhere else?

**Reply by DiegoFleitas** [reply]
> I see. I didn't know that it will not work with tv shows. I will use the updated script but just to be sure, the updated script also works for only movies right?

Correct, unless I get really really bored and decide to code it myself 😄 

> And also about the localstorage save, where is the csv file saved actually? Will it be inside my downloads folder of firefox or somewhere else?

The CSV should be saved on your downloads folder, If you can't see it then the script must be failing. I added a note to the README because I just noticed it seems to throw errors when the region is not US, read it here https://github.com/DiegoFleitas/bookmarklet-collection/tree/main/justwatch-watchlist-exporter

**Reply by bagajohny** [reply]
Your script worked like a charm!! I was able to export everything very quickly.
